readthedocs/api/v3/permissions.py

```diff
@@ -29,3 +29,13 @@ def has_permission(self, request, view):
                 return True
 
         return False
+
+
+class IsProjectAdmin(BasePermission):
```

Comment:
This permission seems to be generating something like 6 extra queries which was surprising. However, it is not generating N queries where N is the number of redirects.

Reply:
Well, this is the basic permission check that we need in all the endpoints for user admin of a project. I split it here to make it simpler and have it isolated.

So, if this check is slow, we should take care of it since it's used a on most of the endpoints.

Basically, it does:

- get a Project (from slug in the URL)
- performs a Project.objects.filter(users__in=[user])

I think we can use `.only('id')` in `ProjectQuerySetMixin.has_admin_permission` in the if, to avoid bringing the full objects since we only want to make a simple check.

In my very simple local tests, there is a timing difference. I'll add a commit for this.

Reply:
Also, some of these method are called more than once (at least for the BrowsableAPIRenderer) and returned value won't change in the same request, so we could do something there if we care enough.

Reply:
Overall it probably isn't a big deal either way. Some number of queries is necessary for the permission check but 6 just seemed high. I do wonder if we shouldn't just cache the project object onto the view object since most of our views relate to projects.

Reply:
I think we can optimize it later if needed --I think it's a good option. Although, I don't want to introduce a "buggy cached object" at this point before knowing that everything is working as we expect :)

Reply:
I'm ok with that for now. Let's keep it in mind though.